File: src/hab/auth/credentials.py
"""Credential storage with encryption."""

from __future__ import annotations

import base64
import hashlib
import json
import os
import platform
import uuid
from typing import Any

from cryptography.fernet import Fernet, InvalidToken
from pydantic import BaseModel, Field

from hab.config.paths import get_credentials_path


def _get_machine_id() -> str:
    """Get a machine-specific identifier for key derivation.

    Returns:
        A string that is unique to this machine.
    """
    # Combine multiple sources for a stable machine identifier
    sources = [
        platform.node(),  # Hostname
        str(uuid.getnode()),  # MAC address
    ]

    # Try to get additional system-specific identifiers
    try:
        if platform.system() == "Darwin":
            # macOS: use hardware UUID
            import subprocess

            result = subprocess.run(
                ["ioreg", "-rd1", "-c", "IOPlatformExpertDevice"],
                capture_output=True,
                text=True,
                timeout=5,
            )
            if result.returncode == 0:
                for line in result.stdout.split("\n"):
                    if "IOPlatformUUID" in line:
                        sources.append(line.split('"')[-2])
                        break
    except (OSError, subprocess.TimeoutExpired):
        pass

    return "|".join(sources)


def _derive_key() -> bytes:
    """Derive an encryption key from machine-specific data.

    Returns:
        A 32-byte key suitable for Fernet encryption.
    """
    machine_id = _get_machine_id()
    # Use SHA-256 to derive a consistent key
    key_material = hashlib.sha256(machine_id.encode()).digest()
    # Fernet requires base64-encoded 32-byte key
    return base64.urlsafe_b64encode(key_material)


class Credentials(BaseModel):
    """Stored credentials for Home Assistant authentication."""

    url: str = Field(description="Home Assistant URL")
    access_token: str | None = Field(default=None, description="Long-lived access token")
    refresh_token: str | None = Field(default=None, description="OAuth refresh token")
    token_expiry: float | None = Field(default=None, description="Token expiry timestamp")

    @property
    def is_oauth(self) -> bool:
        """Check if using OAuth authentication."""
        return self.refresh_token is not None

    @property
    def has_valid_token(self) -> bool:
        """Check if there is a valid access token."""
        return self.access_token is not None


def load_credentials(config_dir: str | None = None) -> Credentials | None:
    """Load credentials from encrypted storage.

    Args:
        config_dir: Optional custom configuration directory.

    Returns:
        Credentials if found and decrypted successfully, None otherwise.
    """
    # First check environment variables
    env_url = os.environ.get("HAB_URL")
    env_token = os.environ.get("HAB_TOKEN")

    if env_url and env_token:
        return Credentials(url=env_url, access_token=env_token)

    # Check for refresh token in environment
    env_refresh = os.environ.get("HAB_REFRESH_TOKEN")
    if env_url and env_refresh:
        return Credentials(url=env_url, refresh_token=env_refresh)

    # Load from file
    creds_path = get_credentials_path(config_dir)
    if not creds_path.exists():
        return None

    try:
        encrypted_data = creds_path.read_bytes()
        fernet = Fernet(_derive_key())
        decrypted = fernet.decrypt(encrypted_data)
        data = json.loads(decrypted.decode())
        return Credentials(**data)
    except (InvalidToken, json.JSONDecodeError, OSError):
        return None


def save_credentials(credentials: Credentials, config_dir: str | None = None) -> None:
    """Save credentials to encrypted storage.

    Args:
        credentials: Credentials to save.
        config_dir: Optional custom configuration directory.
    """
    creds_path = get_credentials_path(config_dir)
    creds_path.parent.mkdir(parents=True, exist_ok=True)

    # Encrypt the credentials
    fernet = Fernet(_derive_key())
    data = json.dumps(credentials.model_dump()).encode()
    encrypted = fernet.encrypt(data)

    # Write with restrictive permissions
    creds_path.write_bytes(encrypted)
    os.chmod(creds_path, 0o600)


def delete_credentials(config_dir: str | None = None) -> bool:
    """Delete stored credentials.

    Args:
        config_dir: Optional custom configuration directory.

    Returns:
        True if credentials were deleted, False if they didn't exist.
    """
    creds_path = get_credentials_path(config_dir)
    if creds_path.exists():
        creds_path.unlink()
        return True
    return False

File: auth/credentials.go
package auth

import (
	"encoding/json"
	"os"
	"time"

	"github.com/home-assistant/hab/config"
)

// Credentials stores authentication information
type Credentials struct {
	URL          string  `json:"url"`
	AccessToken  string  `json:"access_token,omitempty"`
	RefreshToken string  `json:"refresh_token,omitempty"`
	TokenExpiry  float64 `json:"token_expiry,omitempty"`
}

// IsOAuth returns true if using OAuth authentication
func (c *Credentials) IsOAuth() bool {
	return c.RefreshToken != ""
}

// HasValidToken returns true if there is an access token
func (c *Credentials) HasValidToken() bool {
	return c.AccessToken != ""
}

// IsExpired returns true if the token is expired
func (c *Credentials) IsExpired() bool {
	if !c.IsOAuth() {
		return false // Long-lived tokens don't expire
	}
	if c.TokenExpiry == 0 {
		return true // No expiry info, assume expired
	}
	return time.Now().Unix() >= int64(c.TokenExpiry)
}

// NeedsRefresh returns true if the token needs to be refreshed
func (c *Credentials) NeedsRefresh() bool {
	if !c.IsOAuth() {
		return false
	}
	if c.TokenExpiry == 0 {
		return true
	}
	// Refresh if within 5 minutes of expiry
	return float64(time.Now().Unix()) >= (c.TokenExpiry - 300)
}

// LoadCredentials loads credentials from storage
func LoadCredentials(configDir string) (*Credentials, error) {
	// First check environment variables
	envURL := os.Getenv("HAB_URL")
	envToken := os.Getenv("HAB_TOKEN")

	if envURL != "" && envToken != "" {
		return &Credentials{
			URL:         envURL,
			AccessToken: envToken,
		}, nil
	}

	// Check for refresh token in environment
	envRefresh := os.Getenv("HAB_REFRESH_TOKEN")
	if envURL != "" && envRefresh != "" {
		return &Credentials{
			URL:          envURL,
			RefreshToken: envRefresh,
		}, nil
	}

	// Load from encrypted file
	credsPath := config.GetCredentialsPath(configDir)

	data, err := os.ReadFile(credsPath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}

	// Decrypt
	key := deriveKey()
	decrypted, err := decrypt(data, key)
	if err != nil {
		return nil, err
	}

	var creds Credentials
	if err := json.Unmarshal(decrypted, &creds); err != nil {
		return nil, err
	}

	return &creds, nil
}

// SaveCredentials saves credentials to encrypted storage
func SaveCredentials(creds *Credentials, configDir string) error {
	if err := config.EnsureConfigDir(configDir); err != nil {
		return err
	}

	data, err := json.Marshal(creds)
	if err != nil {
		return err
	}

	key := deriveKey()
	encrypted, err := encrypt(data, key)
	if err != nil {
		return err
	}

	credsPath := config.GetCredentialsPath(configDir)
	if err := os.WriteFile(credsPath, encrypted, 0600); err != nil {
		return err
	}

	return nil
}

// DeleteCredentials removes stored credentials
func DeleteCredentials(configDir string) bool {
	credsPath := config.GetCredentialsPath(configDir)
	if err := os.Remove(credsPath); err != nil {
		return false
	}
	return true
}
